rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,294 @@
+
+### Problem resolution
+
+On the nixpkgs configuration side, the first iteration used a generic "predicate" system for ignoring packages, similar to `allowUnfreePredicate` and `allowInsecurePredicate`. This turned out to be both too flexible and not convenient enough to use, so this was complemented with a list of packages to ignore and "smart" default values generation.
+
+A second approach used a list type: `list of ("packageName.warningKind" or "packageName.*" or "*.warningKind" or "*.*")`. This was a binary choice (compared to the ternary value today), with an additional boolean `traceIgnoredWarnings` option. One downside is that it does not allow granular control over warnings, only evaluation failures. A bigger issue is that due to how the merge rules on lists work, it would have been difficult to provide good default values for the nixpkgs confinguration while keeping backwards compatibility.
+
+A third approach used a two-level attribute set, similarly to the list above, but with the value setting the handler instead of it being a binary choice. `"*"."*" = "error";`, `myPackage."*" = "ignore";`, `"*".maintainerless = "ignore";`, etc. This provides better merging behavior than a list, while also being more granular. However, people voiced concerns about people accidentally wildcard-ignoring everything. Also, it is ambiguous on matching problem name vs problem kind, which mostly works fine but might lead to weird things happening in the case where new kinds are introduced.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- ~~From above: "Ignoring a package without issues (i.e. they have all been resolved) results in a warning at evaluation time". How could this be implemented, and efficiently?~~
+  - ~~More generally, how do we tell users that their ignored warning can be removed, so that they won't accidentally miss future warnings?~~
+  - ~~Issues have a `resolved` attribute that may be used for that purpose.~~
+    - Properly implementing this turned out to be non-trivial, so this feature was cut for the sake of simplicity as it was not of high importance anyways.
+  - The ignore mechanism has been refined so that there is less risk of missing future warnings.
+- ~~Should issues be a list or an attrset?~~·
+  - ~~We are using a list for now, there is always the possibility to also allow attrsets in the future.~~
+  - Current design uses an attribute set
+- Currently, many of the relevant nixpkgs configuration options can also be set impurely via environment variables. The `config.problemHandler` option does however not easily map to some environment variable.
+  - When merging existing features into the problems system, existing environment variable will keep working in the future.
+  - Maybe using less environment variables is all for the better?
+  - We may always add specific environment variables for specific use cases where needed without having to expose the full expression power of the configuration options.
```

Comment:
I consider environment variables to overwrite the "refusing of evaluation" quite important.
For example, i quite often install packages with `nix profile` (previously with `nix-env`), because it is faster and i don't want to edit my NixOS configuration for these minor package additions. I also use `allowUnfreePredicate` to only enable some unfree Packages. Sometimes the need arises to install an unfree package with `nix profile`, so i do 
```
env NIXPKGS_ALLOW_UNFREE=1 nix profile install --impure nixpkgs#hello-unfree
```

I haven't done that for broken packages, yet, but i can imagine it being an inconvenience if it wasn't possible for the issue handlers to be overwritten impurely.

Reply:
Just to be clear: Existing environment variables will continue working.

Reply:
At the same time, they will be _the only_ way to mass-override `unsupported`?

Reply:
I'm not sure I understand your question. Why the only way?

Reply:
Can you lower from `error` to `warn` via `matchers`? I thought that `handlers` need to be package-by-package, and `matchers` cannot lower severity below the default?

(Presumably if `unsupported` replaces `platforms` it defaults to `error`)

Reply:
That's a good point. Arguably problems like `unsupported` or `broken` should be safe to ignore generally, while `insecure` ones aren't..

Maybe the distinction here is that `unsupported` and `broken` are errors at evaluation time that would otherwise happen at build/runtime, while `insecure` wouldn't give any other error.

Reply:
I've personally almost never interacted with `unsupported` and only rarely with `broken`, do you think that not being able to mass-disable them is a big issue?

On the other hand, my previous design for this RFC included some pretty YOLO wildcard capabilities, so I'll have no problems with selectively relaxing the "no downgrade" restriction should there be need for it.

Also, for better or for worse, the existing environment variables will keep working for a long time due to backwards compatibility. It's a bit weird, but there's nothing we can really do about it.

Reply:
Maybe some thing can be downgraded to warn but not to nothing.

The use case for mass-overriding unsupported/broken/etc. is e.g. «so with the new and improved stdenv for this kind of BSD which has partial Linux compatibility, which of the presumed-Linux-only-packages will now build?»

Sure, on non-cross Linux `broken` usually is in some connection to reality; for new or reworked exotic stdenvs (Tier 3 or 4 platforms, including cross or maybe musl or whatever) «there were reasons to guess Linux-only or x86_64-only, and nobody had a setup to try» is pretty normal and quickly estimating the amount of work to get something working is useful.